### DISCUSSION TITLE
Expose `renderToString` method

### DISCUSSION THREAD

**Discussion by waspeer** [withastro/roadmap#409] [Proposal]
- Start Date: 2022-12-07
- Reference Issues: https://github.com/withastro/astro/issues/4372 https://github.com/withastro/astro/issues/5494 
- Implementation PR: 

# Summary

Expose a method to render an astro component to a string.

# Example

```astro
---
// In an Astro component
import Component from './Component.astro';
const html = Astro.renderToString(<Component name="world" />);
---
```

and/or

```ts
// (Possibly) in a TypeScript file
import { renderToString } from 'astro';
const html = renderToString('./Component.astro', { props: { name: 'world' } });
```

# Background & Motivation

Most of the use cases of rendering components are covered with the features of Astro sfc's. However, there are cases where you'd want to 'break out' of the sfc, and render an Astro component to a string.

I'll show two use cases; one within an Astro component, and one within an endpoint.

## Within an Astro component

[Sanity CMS](https://www.sanity.io/) has a custom format for rich text called [Portable Text](https://github.com/portabletext/portabletext). To render this they provide [a couple of packages](https://www.npmjs.com/search?q=%40portabletext), of which I'll be using [`@portabletext/to-html`](https://github.com/portabletext/to-html).

The gist of Portable Text is that it's an array of `blocks`  that render differently based on their properties. There are preconfigured block types (eg `h1`, `h2`, `p`), but it's also possible to define custom blocks, which you'd render like this:

```astro
---
import SanityImage from './SanityImage.astro';
import CallToAction from './CallToAction.astro';
import CallToActionInline from './CallToActionInline.astro';

import { toHTML } from '@portabletext/to-html';

const { somePtValue } = Astro.props;
---

<Fragment set:html={
  toHTML(somePtValue, {
    components: {
      types: {
        image: ({value}) => Astro.renderToString(<SanityImage asset={value.asset} />),
        callToAction: ({value, isInline}) =>
          isInline
            ? Astro.renderToString(<CallToActionInline {...value} />)
            : Astro.renderToString(<CallToAction {...value} />),
      },
    }
  });
} />
```

## Within an endpoint

```ts
// src/pages/files/[fileId]/download.ts
import type { APIRoute } from 'astro';
import { getFileInfoById } from '~/lib/files';

export const get: APIRoute = async ({ params }) => {
  const { fileId } = params;

  const fileInfo = await getFileInfoById(fileId);

  if (!fileInfo) {
    return new Response(
      renderToString('./components/404.astro', { 
        props: { 
          message: `File with id ${fileId} not found`, 
        },
      }),
      { 
        status: 404,
      }
    );
  }

  // ... //
}
```

# Unresolved Questions

I do not have a clear idea yet about what this means for CSS and JS included in the component that's rendered to string. My first intuition is to include the CSS and JS in the string, but I'm not sure if that's the best approach. 

Another option could be to return the HTML, CSS and JS separately, but that would complicate the solutions to both the provided use cases above.

I'd love to hear your thoughts on this.

**Comment by RealOrangeOne** [comment]
This is one of the biggest misses in Astro IMO. Writing an endpoint, being able to list pages, but not render any of their content is very annoying. Even inside a component, trying to render say a summary of page content for use in SEO metatags is almost impossible without resorting to using a separate Astro component to re-render the slot, convert the HTML to text then truncate it (which doesn't quite work due to some [other issues](https://github.com/withastro/astro/issues/6220)).

The embedded stylesheets / JS is an interesting one, so perhaps much like `.render` works at the moment, return multiple values. The content as a HTML string, plus some additional references to the rendered stylesheets to embed. Alternatively, strip those out and just return the text, as I suspect that's where a large amount of the value comes from.

**Comment by Princesseuh** [comment]
This is covered by https://github.com/withastro/roadmap/issues/533, so closing this discussion. Thank you for the suggestion! For further feedback, please post there.
